src/main/ts/core/Changelog.ts

```diff
@@ -0,0 +1,361 @@
+      if (seenLF) {
+        // add line start and reset
+        lines.push(i);
+        seenCR = false;
+        seenLF = false;
+      }
+      seenLF = true;
+    } else if (c === '\r') {
+      if (seenCR) {
+        // add line start and reset
+        lines.push(i);
+        seenCR = false;
+        seenLF = false;
+      }
+      seenCR = true;
+    } else {
+      if (seenLF || seenCR) {
+        // add line start and reset
+        lines.push(i);
+        seenLF = false;
+        seenCR = false;
+      }
+    }
+  }
+  return { text, lines };
+};
+
+const posToOffset = (lines: number[], position: [number, number]) => {
+  // both line and column count from 1
+  const [ line, column ] = position;
+  return lines[line - 1] + (column - 1);
+};
+
+// the range including just the columns of the node
+const columnRange = (lines: number[], node: commonmark.Node): Offset => {
+  // this is the inclusive range
+  const [ startPos, endPos ] = node.sourcepos;
+  const start = posToOffset(lines, startPos);
+  const end = posToOffset(lines, endPos);
+  return { start, end };
+};
+
+// the range including the whole lines that the node is in
+const blockRange = (source: TextLines, node: commonmark.Node): Offset => {
+  const [[ startLine ], [ endLine ]] = node.sourcepos;
+  const start = posToOffset(source.lines, [ startLine, 1 ]);
+  const end = (endLine < source.lines.length ? posToOffset(source.lines, [ endLine + 1, 1 ]) : source.text.length) - 1;
+  return { start, end };
+};
+
+const extractText = (source: TextLines, node: commonmark.Node): string => {
+  const range = columnRange(source.lines, node);
+  return source.text.substring(range.start, range.end + 1);
+};
+
+const pos = (node: commonmark.Node): string => {
+  if (node.sourcepos !== undefined) {
+    const [[ startLine, startColumn ]] = node.sourcepos;
+    return ` (line: ${startLine} column: ${startColumn})`;
+  } else {
+    return '';
+  }
+};
+
+const parseHeadings = (changelog: string): Top => {
+  const data: Top = { preamble: [], subheadings: [] };
+  const reader = new commonmark.Parser();
+  const doc = reader.parse(changelog);
+  const stack: (Heading | Top)[] = [ data ];
+  let node: commonmark.Node | null = doc.firstChild;
+  while (node !== null) {
+    if (node.type === 'heading') {
+      // pop headings off the stack that are higher or equal level
+      while (true) {
+        const last = stack[stack.length - 1];
+        if (node.level <= (last?.header?.level ?? 0)) {
+          stack.pop();
+        } else {
+          break;
+        }
+      }
+      // push this as a new child heading
+      const heading: Heading = {
+        header: node,
+        preamble: [],
+        subheadings: []
+      };
+      stack[stack.length - 1].subheadings.push(heading);
+      stack.push(heading);
+    } else {
+      stack[stack.length - 1].preamble.push(node);
+    }
+    node = node.next;
+  }
+  return data;
+};
+
+const jiraRe = /\s+#(?<jira>[A-Z]{2,10}-\d+)\s*$/;
```

Comment:
We should name this more generically, as this won't always be a Jira ticket number

Reply:
(By that I mean references to "Jira" in here, not just the regex)

Reply:
Suggestions?

Reply:
I use the term "ticket", I've also seen "issue". For TinyMCE these can be JIRA numbers or GitHub issue numbers and for the latter we do in the changelog `#GH-XXXX`, so this regex will still work.

Reply:
Ok, will rename to "ticket".